006-strimzi-ui.md

```diff
@@ -0,0 +1,129 @@
+
+### UI implementation
+
+At a high level, I would propose a UI implemented as follows:
+
+- It would be a Javascript based UI (using [babel](https://babeljs.io/) to provide latest ECMAScript capabilities in a cross browser compatible manner), using the [React](https://reactjs.org/) framework
+- This UI is hosted and provided via an [Express](https://expressjs.com/) server
+- This UI is built in a modular/metadata driven manner - allowing for easy extensibility, modification, and dynamic behaviour at runtime
+- Uses a [GraphQL](https://graphql.org/) API
+- Should be fully translatable, and accessible
+
+I would suggest that the UI implementation to follow the [Model View Controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) (MVC) pattern, with clear separation between business logic, and the view logic that renders the state of the UI. This would enable/allow the view logic to be modified/swapped in/out as required, but keep the same business logic. I would suggest using the [Carbon](https://www.carbondesignsystem.com/) design system as the view layer for the contributed UI, given it is open source, is supported, and I can provide both the design and implementation using it. As an example, the following is a mock up of how a Topics page could look using the Carbon design system, which would allow a user to view, create, edit and delete topics in a given cluster:
+
+![Topics mock up](./006-topicsdesign.png)
+
+By maintaining and enforcing a thin view layer, this should not stop other view layers being designed, implemented and contributed in other frameworks, such as [PatternFly](https://www.patternfly.org/v4/) or [Material](https://material.io/) for example.
+
+I would also suggest using Express as a server for this UI. This is due to its small footprint, modularity and available 3rd party modules, such as [helmet for http security](https://helmetjs.github.io/), [Passport.js for authentication and authorization support](http://www.passportjs.org/) or [the graphql and express-graphql modules for GraphQL server support](https://graphql.org/graphql-js/running-an-express-graphql-server/).
+
+This UI would also be provided with a full set of supporting elements - such as end to end tests, automation (examples of which are [in this section](#a-ui-repository-and-ways-of-working-in-the-repository)) to manage common tasks, and documentation around implementation approach, UI best practise and so on.
+
+The below shows how these pieces could integrate.
+
+![Suggested topology](./006-topology.png)
+
+Where:
+
+- 'strimzi-api' is a backend server for Kafka data/requests
+- 'Supporting elements' could be other Kafka related deployments such as Mirror Maker, Kafka Connect and Cruise control. These could also be non Kafka related deployments (not deployed as a part of Strimzi), such as LDAP servers, metric stores etc.
```

Comment:
Yes, that's the idea. I guess that from a JavaScript UI client perspective is always better/simple interacting with a REST HTTP or GraphQL API than a Kubernetes API server (even if it's HTTP at the end). It means that we are not going to rely on Strimzi custom resource for getting cluster and topic information but using the Kafka Admin client in the backend.
Do you see any specific reason why we could need some interaction with the Kubernetes API?

Reply:
It will interact with a backend for Kafka data - hence the explicit mentioning of it. Given the discussion it will be the implementation discussed in strimzi/strimzi-admin#4 I imagine. And as required, other data backend APIs could be added, using the UI server as a layer of abstraction/configuration point to enable them.

Reply:
And to your comment @ppatierno - it might be interesting for example for some degree of pod status of a deployed Strimzi instance to be surfaced in a 'live status' type UI (eg X of Y Kafka pods running, Z of Y crashloopbackoff etc). At which point the UI of course would need a service to interact with kube to get pod status.